Add telemetry to System.Net.Http

davidsh opened the pull request:
This is the first of several PRs to add telemetry focused events and counters.

A few things to note:

* Added a new dedicated namespace for telemetry. Current feedback from partners is that our
existing events are "confusing and a mess" due to their focus on low-level diagnostics and verbose
debugging. See discussion in #37428 regarding reasons for a new namespace.

* Please comment on the naming patterns for the events, counters, and
display names, etc. I looked at ASP.NET events/counters and @anurse older guidance in
https://gist.github.com/anurse/af1859663ac91c6cf69c820cebe92303. But there are inconsistencies
between all of that. So, I'd like to get agreement on naming patterns for events/counters
(i.e. noun-verb vs. verb-noun, present tense vs. past tense, etc.).

* No automated tests were added at this time. The current discussions with others (reverse proxy
team) is that automated tests are brittle in CI. But manual tests will be run.

Contributes to #37428

davidsh committed:
[WIP] Add telemetry to System.Net.Http

This is the first of several PRs to add telemetry focused events and counters.

A few things to note:

* Added a new dedicated namespace for telemetry. Current feedback from partners is that our
existing events are "confusing and a mess" due to their focus on low-level diagnostics and verbose
debugging. See discussion in #37428 regarding reasons for a new namespace.

* Please comment on the naming patterns for the events, counters, and
display names, etc. I looked at ASP.NET events/counters and @anurse older guidance in
https://gist.github.com/anurse/af1859663ac91c6cf69c820cebe92303. But there are inconsistencies
between all of that. So, I'd like to get agreement on naming patterns for events/counters
(i.e. noun-verb vs. verb-noun, present tense vs. past tense, etc.).

* No automated tests were added at this time. The current discussions with others (reverse proxy
team) is that automated tests are brittle in CI. But manual tests will be run.

Contributes to #37428

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,104 @@
+                // The cumulative number of HTTP requests started since the process started.
+                _startedRequestsCounter ??= new PollingCounter("requests-started", this, () => _startedRequests)
+                {
+                    DisplayName = "Requests Started",
```
I think the display names should be in resources?

davidsh replied:
I followed this pattern from ASP.NET. They don't use resources for these strings. It is similar perhaps to other logging where we do not localize at all.

@davidfowl @Tratcher Can you comment?

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,104 @@
+                // They aren't disabled afterwards...
+
+                // The cumulative number of HTTP requests started since the process started.
+                _startedRequestsCounter ??= new PollingCounter("requests-started", this, () => _startedRequests)
```
I'm not familiar with how these event counters are exposed -- I assume they're scoped under "System.Net.Http" (the `EventSource(Name=...)' attribute above)?

davidfowl replied:
Yep

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,104 @@
+                // They aren't disabled afterwards...
+
+                // The cumulative number of HTTP requests started since the process started.
+                _startedRequestsCounter ??= new PollingCounter("requests-started", this, () => _startedRequests)
```
The delegates should be using `Interlocked.Read` to read `_startedRequests` etc. avoid tearing/visibility issues.

davidsh replied:
I followed this pattern from ASP.NET. They don't use Interlocked.Read() for these.

@davidfowl @Tratcher Can you comment?

davidfowl replied:
Hmm maybe we're tearing on 32 bit! 😄

davidsh replied:
I opened an issue for this: https://github.com/dotnet/aspnetcore/issues/22888

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -345,7 +345,25 @@ private HttpMessageHandler SetupHandlerChain()
+                // Wrap the task for event activity-id threading.
+                HttpTelemetry.Log.RequestStart(request.RequestUri.IdnHost, request.RequestUri.Port);
+                Task<HttpResponseMessage> task = handler.SendAsync(request, cancellationToken);
+                task.ContinueWith(
+                    t => HttpTelemetry.Log.RequestAbort(),
+                    cancellationToken,
+                    TaskContinuationOptions.OnlyOnCanceled | TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+                task.ContinueWith(
+                    t => HttpTelemetry.Log.RequestStop(request.RequestUri.IdnHost, request.RequestUri.Port),
+                    cancellationToken,
+                    TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```
The cancellation token provided to `ContinueWith` will cancel the continuation from running, which will cause the initial `RequestStart` to not see a corresponding end.

Also consider passing in the `request` in for `Object`-taking overload, to prevent allocating a per-request closure.

It might also be faster to return the continuation task instead of the original task. I believe @stephentoub made a change that allows only a single continuation to run synchronously, so when the user `await`s the original `Task` it will always end up punted to the scheduler due to your continuation getting synchronous priority.

stephentoub replied:
We should combine both continuations into a single async method, e.g.:
```C#
return HttpTelemetry.IsEnabled && request.RequestUri != null ?
    WithLogging(handler, request, cancellationToken) :
    handler.SendAsync(request, cancellationToken);

static async Task WithLogging(HttpMessageHandler handler, HttpRequestMessage request, CancellationToken cancellationToken)
{
    HttpTelemetry.Log.RequestStart(request.RequestUri.IdnHost, request.RequestUri.Port);
    try
    {
        await handler.SendAsync(request, cancellationToken).ConfigureAwait(false);
    }
    catch
    {
        HttpTelemetry.Log.RequestAbort(); // this doesn't need the request passed in?
        throw;
    }
    finally
    {
        HttpTelemetry.Log.RequestStop(request.RequestUri.IdnHost, request.RequestUri.Port);
    }
}
```

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -345,7 +345,25 @@ private HttpMessageHandler SetupHandlerChain()
             }
 
-            return handler.SendAsync(request, cancellationToken);
+            if (HttpTelemetry.IsEnabled && request.RequestUri != null)
```
Is there a case where `RequestUri` would be `null` here?

davidsh replied:
Probably not but request.RequestUri is a property that can return null in theory. I was getting nullable compile errors without this check. 

@stephentoub Can you comment on this?

stephentoub replied:
It can be null here, e.g. it'll be null here if you do:
```C#
var invoker = new HttpMessageInvoker(new SocketsHttpHandler());
await invoker.SendAsync(new HttpRequestMessage(), default);
```

stephentoub replied:
(I don't know why ValidateAndNormalizeRequest doesn't produce an exception if the Uri is null.  If we wanted to change that so that it did and we'd never get here with a null Uri, then we could Debug.Assert(request.RequestUri != null) instead.)

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,104 @@
+                };
+
+                // The number of HTTP requests aborted per second since the process started.
+                _requestsPerSecondCounter ??= new IncrementingPollingCounter("requests-aborted-per-second", this, () => _abortedRequests)
```
These per-second counters can be derived from their absolute variants. Do these take extra CPU to log?

davidsh replied:
@noahfalk Can you also comment on this? I think the overhead of this counter is small.

noahfalk replied:
The performance overhead of a polling counter should be extremely low and it scales based on how frequently the telemetry listener wants to poll it. What will happen is every N seconds (the telemetry listener chooses N) your callback function will be invoked, whatever value it returns will be serialized into a message and that message will be emitted into the event stream. Assuming your callback is not CPU intensive this should translate into a few allocations and maybe maybe 1us of CPU activity every N seconds. 
Likely the biggest performance cost to be aware of is the cost the library pays to update the _abortedRequests field on each request, but you are fully in control of that overhead.

davidsh replied:
Thanks @noahfalk for the feedback. Based on this I'll leave these separate counters.

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,104 @@
+// TODO: Get rid of this once I figure out how to handle build errors like these given the current lazy creation pattern:
+//     System\Net\Http\HttpTelemetry.cs(19,32): error CS8618: Non-nullable field '_startedRequestsCounter' is uninitialized.
+//     Consider declaring the field as nullable. [S:\GitHub\runtime\src\libraries\System.Net.Http\src\System.Net.Http.csproj]
+#nullable disable
```
You can do `_startedRequestsCounter = null!` if you know they'll be initialized later prior to use.

davidsh replied:
I tried that. But then I got compile errors from the nullable checks. Basically I didn't want to declare these fields as nullable. The logic is such that they are not null when not being used.

@stephentoub Is there a way to handle this in the logic without making the fields nullable?

stephentoub replied:
_startedRequestsCounter should be nullable:
```C#
private PollingCounter? _startedRequestsCounter;
```
What problems does that cause?

campersau commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -345,7 +345,25 @@ private HttpMessageHandler SetupHandlerChain()
+            {
+                // Wrap the task for event activity-id threading.
+                HttpTelemetry.Log.RequestStart(request.RequestUri.IdnHost, request.RequestUri.Port);
+                Task<HttpResponseMessage> task = handler.SendAsync(request, cancellationToken);
```
`SendAsync` might throw synchronously and then no `RequestStop` would be logged.
For example when the `IWebProxy` implementation throws or for example here: https://github.com/dotnet/runtime/blob/2b7852b3d039b7d40f2623a3ec54e7d23eb93790/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L379

halter73 commented:
> No automated tests were added at this time. The current discussions with others (reverse proxy team) is that automated tests are brittle in CI. But manual tests will be run.

What makes testing this brittle?

I don't think Kestrel's EventSourceTests have been flaky in the past year an a half. The last time I remember flakiness in those tests, we determined it was caused by threadpool starvation which in turn was caused by sync-over-async when disposing Kestrel's test server. https://github.com/dotnet/aspnetcore/issues/6127

davidsh committed:
Address PR feedback

* PR feedback
* Added RequestAbort as separate event method with dimensions (host, port)

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -345,7 +346,28 @@ private HttpMessageHandler SetupHandlerChain()
+                }
+                finally
+                {
+                    HttpTelemetry.Log.RequestStop(request.RequestUri.IdnHost, request.RequestUri.Port);
```
I brought up [in another issue](https://github.com/dotnet/runtime/issues/37141) that Http telemetry appears to count receiving a message's headers as the request being done, not when the request is finished.

Is RequestStop the right name for this?

Is current requests the right counter when a request can still be in-progress?

davidsh replied:
The 'Start' and 'Stop' suffixes are required in order for the EventSource activity threading to work automatically.  See comments in source code in the PR.

As discussed in #37141, HttpClient and the handler architecture define a request as 'done' when the request headers arrive.  Download of the response body is outside of the handlers themselves.

stephentoub replied:
> As discussed in #37141, HttpClient and the handler architecture define a request as 'done' when the request headers arrive. Download of the response body is outside of the handlers themselves.

We could make it work if we wanted to, albeit at an expense.  We could wrap the returned stream in a custom one that passed through the calls to the wrapped stream, and then signaled Stop when EOF is hit.

davidsh replied:
>We could make it work if we wanted to

Do we want to? This is an interesting idea. We would have to make it work across HTTP versions (1.x, 2, 3).

Perhaps I can file a separate issue for this. For now, we can do it as I've done in the PR.  But use the separate issue to further discuss this topic and perhaps revise accordingly?

stephentoub replied:
Separate issue sounds fine. thanks.

Tratcher replied:
The connection pool could probably better track active requests end-to-end by tracking http/1.x connections and h2 streams in use. Tracking stop and current-requests at the SocketHttpHandler level gives confusing results.

davidsh replied:
I added it to SocketsHttpHandler because it maps better to how the public SendAsync API works and it avoids having to add it in multiple places for HTTP 1.x, 2, 3 versions.

>Tracking stop and current-requests at the SocketHttpHandler level gives confusing results.

I don't understand why you consider it confusing.

Tratcher replied:
> HttpClient and the handler architecture define a request as 'done' when the request headers arrive. Download of the response body is outside of the handlers themselves.

HttpClient buffers most responses by default, those aren't considered done until fully buffered.

The naming/namespace of RequestStart/Stop and current-requests does not specify that these are scoped to the HttpClient or SocketHttpHandler. They sound like end-to-end lifetime events. Maybe the EventSource name for these should be changed to System.Net.Http.SocketHttpHandler.

scalablecory replied:
I agree with @Tratcher. We have following scenarios:

- SendAsync unbuffered
- SendAsync buffered
- Convenience methods (buffered)

It will be confusing for the latter two if the timings don't measure the entire operation. We should consider separate "headers received" and "request 100% finished" timings.

davidsh replied:
>We should consider separate "headers received" and "request 100% finished" timings.

I agree we can add separate events for things like "headers received" and "response body fully read". Those can be sprinkled into SocketsHttpHandler (we already have diagnostic events at similar points).

But we still need to provide a definitive 'Stop' event for activity threading. So, are we thinking that we want 'RequestStop' to always mean response headers read and response body download complete?

stephentoub replied:
> But we still need to provide a definitive 'Stop' event for activity threading. So, are we thinking that we want 'RequestStop' to always mean response headers read and response body download complete?

From a logging / counters perspective, I'm not sure that having received response headers is particularly relevant.  Seems by default we should be tracking the whole request/response; if there's an advanced scenario that wants to be notified of just headers having been received, that could be done, too, in some fashion.

wfurt replied:
Even if we do not buffer response, the stream we create does know when the response is done, right? (and we try to drain if we hand out stream and that is not read to the end)

davidsh replied:
Based on this discussion and the consensus of everyone involved, I will change the PR so that 'RequestStop' includes reading the response body until EOF (or error).

I will also re-activate #37141. But that issue was about the DiagnosticSource telemetry in DiagnosticHandler.

JamesNK commented:
An example of unit testing counters from grpc-dotnet: https://github.com/grpc/grpc-dotnet/blob/486446086540381be32d40dd675d47da6f989dff/test/FunctionalTests/Client/EventSourceTests.cs

They were a pain in the butt to get right, but they're not flaky.

davidsh commented:
>They were a pain in the butt to get right, but they're not flaky.

Thanks for the pointer to those tests. Some networking event sources (Microsoft-System-Net) diagnostic tracing have some unit test coverage as well.

I agree that with effort automated tests can be written. The topic we discussed at an earlier team meeting was about "is it worth the dev cost right now". I'll add a tracking issue for adding automated test coverage for events/counters where we have gaps.

Tratcher commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -345,7 +346,28 @@ private HttpMessageHandler SetupHandlerChain()
+            static async Task<HttpResponseMessage> WithLogging(HttpMessageHandler handler, HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                Debug.Assert(request.RequestUri != null);
+                HttpTelemetry.Log.RequestStart(request.RequestUri.IdnHost, request.RequestUri.Port);
```
Are these really the interesting fields? Yes Host is interesting, but I'd expect Scheme, Path, and protocol version to be more interesting than Port.

scalablecory replied:
How reasonable would it be to just log the full URI?

davidsh replied:
EventSource events need simple fields. We can't use a complex type like System.Uri.

DiagnosticSource (like the DiagnosticHandler we have) can handle complex objects. But this telemetry is about EventSource/EventCounter.

Tratcher replied:
Uri has an AbsoluteUri string field that could be logged, or we can log some of the more interesting components like Path.

davidsh replied:
That's fine. I was just responding to @scalablecory comment about the "full URI" since EventSource methods (events) need to have simpler arguments since they get emited that way in the traces.

davidsh replied:
I pushed a commit to add more fields to the 'RequestStart' event. I also removed the fields from the RequestStop since RequestStart/RequestStop form an activity pair and the data can be inferred. This also helps makes the events more concise for perf considerations.

davidsh committed:
Address PR feedback 2

* RequestStart now includes host, port, scheme, path and version
* RequestStart and RequestStop now bracket the lifetime of the request including reading the
response body stream. If the response body is not complete read then disposing the stream
will trigger RequestStop as well.

noahfalk commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        [Event(3, Level = EventLevel.Error)]
+        public void RequestAbort()
```
The most common pattern I see (though not by a large margin) seems to be NounVerb where the Verb is written past tense, with the aggravating exception of 'Start' and 'Stop' which are almost always written in that imperative form. IMO past tense also seems more intuitive given that we are almost always recording something that happened in the recent past.

davidsh replied:
I will say that ASP.NET guidelines also adopted NounVerb but used 'present' tense and not 'past' tense:
https://gist.github.com/anurse/af1859663ac91c6cf69c820cebe92303

>Use present-tense verbs (Timeout not TimedOut)
Use NounVerb form (ConnectionStart not StartConnection)

But I tend to agree with you about "past tense also seems more intuitive given that we are almost always recording something that happened in the recent past."

So, I'll change these to use past tense (except for 'Start' and 'Stop').

noahfalk replied:
@davidsh do you know where you found the link to Andrew's gist? It would be nice if we can move the guidance to a location where the team can collectively edit it and redirect whatever link you followed.

davidsh replied:
I found it here originally: https://github.com/dotnet/aspnetcore/issues/4769#issue-390516127

noahfalk replied:
Thanks! I opened PR https://github.com/dotnet/aspnetcore/pull/23036 to move the content of the gist. Once it is moved we'll be able to propose changes via PR and update the old links.

noahfalk commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,100 @@
+        // - A stop event's event id must be next one after its start event.
+
+        [Event(1, Level = EventLevel.Informational)]
+        public void RequestStart(string host, int port, string scheme, string path, string version)
```
Is the version a major/minor pair? There is likely a small perf optimization to pass this as a pair of shorts rather than a string. The string incurs one or more heap allocations, some formatting, a couple buffer copies and a few extra bytes in the final encoding.

davidsh replied:
Good idea. HTTP versions are typically like 1.0, 1.1, 2.0, 3.0. So, this sounds like a good optimization.

noahfalk commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -345,7 +346,30 @@ private HttpMessageHandler SetupHandlerChain()
+                catch
+                {
+                    HttpTelemetry.Log.RequestAbort();
+                    HttpTelemetry.Log.RequestStop();
```
Any risk of this logging stop more than once due to the RequestStop() call in the dispose path? (I didn't scrutinize the http code, I trust you guys are in the best position to decide if this is relevant)

davidsh replied:
Potentially, see https://github.com/dotnet/runtime/pull/37619#discussion_r441164516.

I'll work on fixing this potential issue.

noahfalk commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -345,7 +346,30 @@ private HttpMessageHandler SetupHandlerChain()
+                {
+                    return await handler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                }
+                catch
```
Debugging exception noise is lessened if you use an exception filter rather than catch/rethrow

davidsh replied:
@noahfalk What "noise" are you referring to? In general, we want to catch everything here so that we can guarantee generating the 'RequestAborted' event.

noahfalk replied:
A try/catch/throw pattern throws the exception twice, so the debugger will potentially stop on it twice with a UI popup and log it twice. A little more annoyingly if the debugger is configured only to stop on unhandled exceptions (the default setting) then at best the debugger will stop where the exception is rethrown here but the initial location of the throw will never be considered unhandled because of this catch clause. This in turn means stack frames and useful local variables at the original throw site have already been unwound when the debugger stops later. Async code has similar catch/rethrow behavior so chances are user experience still won't be ideal, but I'd be more optimistic that the debugger can find ways to mitigate the issues caused by async in the future whereas its harder to mitigate explicitly written try/catch/throw clauses with arbitrary logic inside.

Doing a filter shouldn't interfere with you being able to detect all thrown exceptions. If you haven't seen it before the pattern looks like this:
````
try
{
    Console.WriteLine("throw something");
    throw new AnyException();
}
catch (Exception e) when HandleException(e)
{
    Debug.Fail() // this never runs
}

bool HandleException(Exception e)
{
    Console.WriteLine("Exception occured: " + e.GetType());
    return false; // returning false means the catch handler isn't run, the exception isn't considered to be caught so there is no need to throw it again
}
````

davidsh replied:
Thanks! This is a good pattern. I'll incorporate this suggestion.

noahfalk reviewed (commented):
Sorry for the delay @davidsh, it looked pretty good to me and hopefully the comments are useful for you.

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -345,7 +346,30 @@ private HttpMessageHandler SetupHandlerChain()
+                    request.RequestUri.IdnHost,
+                    request.RequestUri.Port,
+                    request.RequestUri.Scheme,
+                    request.RequestUri.AbsolutePath,
```
If perf is important here, `PathAndQuery` could be used instead of `AbsolutePath`, as the string allocation cost for `PathAndQuery` will be incurred by the request anyway.

MihaZupan replied:
Nit:  If parameter order doesn't matter for EventSource, the order of appearance in Uri could be used: `scheme, host, port, path`

davidsh replied:
Thx. I will incorporate these suggestions.

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,100 @@
+        {
+            Interlocked.Increment(ref _startedRequests);
+            Interlocked.Increment(ref _currentRequests);
+            WriteEvent(1, host, port, scheme, path, version);
```
Does this bind to `WriteEvent(id, params object[])`? 

I assume we need a custom overload here to remove the allocations, as done in [NetEventSource](https://github.com/dotnet/runtime/blob/bab2bc4bfaaae52860e8f6f8ef1770c1030e70f1/src/libraries/Common/src/System/Net/Logging/NetEventSource.Common.cs#L505).

davidsh replied:
Yes, I'll add a custom overload and add it to this PR.

In a future PR, we'll refactor all the custom overloads of `WriteEvent` from https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/Net/Logging/NetEventSource.Common.cs and place all of them in a single file.

Tratcher commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs:
```diff
@@ -380,6 +382,8 @@ private void Cancel()
                     _waitSource.SetResult(true);
                 }
+
+                if (HttpTelemetry.IsEnabled) HttpTelemetry.Log.RequestStop();
```
Not Abort?

halter73 replied:
I agree about Abort. It also looks like this is outside of a lock. Does this risk double-counting RequestStop?

davidsh replied:
I've been chasing a few bugs right now with this PR about correct numbers of 'Start' and 'Stop' events. I'll verify this and potentially add some synchronization for `RequestStop` to avoid double-counting.

davidsh commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,100 @@
+        public void RequestStart(string host, int port)
+        {
+            Interlocked.Increment(ref _startedRequests);
+            Interlocked.Increment(ref _currentRequests);
```
Thx. I'll incorporate that suggestion into this PR.

davidsh commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,100 @@
+                };
+
+                // The number of HTTP requests started per second since the process started.
+                _requestsPerSecondCounter ??= new IncrementingPollingCounter("requests-started-per-second", this, () => Interlocked.Read(ref _startedRequests))
```
Thx. I've changed the two counters' names for `requests-started-rate` and `requests-aborted-rate`.

davidsh commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,100 @@
+                };
+
+                // The number of HTTP requests aborted per second since the process started.
+                _requestsPerSecondCounter ??= new IncrementingPollingCounter("requests-aborted-per-second", this, () => Interlocked.Read(ref _abortedRequests))
```
Done. See https://github.com/dotnet/runtime/pull/37619#discussion_r441159647

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,152 @@
+                };
+
+                // The current number of active HTTP requests that have started but not yet completed or aborted.
+                _currentRequestsCounter ??= new PollingCounter("current-requests", this, () => Interlocked.Read(ref _startedRequests) - Interlocked.Read(ref _stoppedRequests))
```
```suggestion
                _currentRequestsCounter ??= new PollingCounter("current-requests", this, () => -Interlocked.Read(ref _stoppedRequests) + Interlocked.Read(ref _startedRequests))
```

`-stopped + started` could be used instead of `started - stopped` to avoid avoid returning a negative value if `_stoppedRequests` is incremented after reading `_startedRequests`

davidsh replied:
It's not possible for _stoppedRequests to be greater than _startedRequests. In most cases with active requests, _stoppedRequests will always be less than _startedRequests. And when all http requests finish completely (response body fully read or it aborts) _stoppedRequests will be equal to _startedRequests.

So, I don't see the value of this change.

noahfalk replied:
I suspect @MihaZupan was noticing that there is a race condition because the read operations for _startedRequests and _stoppedRequests occur at different times. In particular imagine starting with _startedRequests = _stoppedRequests = 0 and then this order of execution:
**Counter polling thread:** Read _startedRequests = 0
**Http Event processing thread:** Complete http request is processed _startedRequests = _stoppedRequests = 1
**Counter polling thread:** Read _stoppedRequests = 1
**Counter polling thread:** Compute 0 - 1 = -1 current requests

davidsh replied:
Thx. I see now.

Doing the math as:

`-Interlocked.Read(ref _stoppedRequests) + Interlocked.Read(ref _startedRequests)`

would end up as:

Counter polling thread: Read _stoppedRequests = 0
Http Event processing thread: Complete http request is processed _startedRequests = _stoppedRequests = 1
Counter polling thread: Read _startedRequests = 1
Counter polling thread: Compute -0 + 1 = 1 current requests

Still not perfectly accurate to transitory race conditions but at least it is a positive number.

I'll make this change.

davidsh commented:
I pushed a new commit that redid how the RequestStart/RequestStop works. I discovered my first attempt at placing the RequestStart into SocketsHttpHandler didn't properly work with redirected requests. We ended up getting extra RequestStop events since more requests are generated internally by AllowAutoDirect=true.

This new commit seems to work good in initial manual testing with HTTP/1.1, HTTP/2.0 requests including redirection. I don't plan to make HTTP/3 work at this time.

davidsh committed:
Address feedback 3

* Use exception filter to improve performance

davidsh committed:
Avoid negative 'current-requests' counter value

davidsh commented:
@dotnet/ncl - Any additional feedback? If not, can someone please sign off?

scalablecory reviewed (approved):
lgtm after these comments.

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,154 @@
+        // - A stop event's event id must be next one after its start event.
+
+        [Event(1, Level = EventLevel.Informational)]
+        public void RequestStart(string scheme, string host, int port, string pathAndQuery, int versionMajor, int versionMinor)
```
Afaik, the method signature has to match the emitted event

davidsh replied:
>Afaik, the method signature has to match the emitted event

Yes, that is the reason.

davidsh committed:
Address feedback 4

davidsh commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -6,6 +6,7 @@
 using System.Net.Security;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Diagnostics;
```
Fixed

davidsh commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs:
```diff
@@ -339,6 +339,44 @@ public Task<HttpResponseMessage> SendProxyConnectAsync(HttpRequestMessage reques
+                throw;
+            }
+
+            bool LogException(Exception e)
```
Fixed

davidsh commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs:
```diff
@@ -0,0 +1,154 @@
+        public void RequestStart(string scheme, string host, int port, string pathAndQuery, int versionMajor, int versionMinor)
+        {
+            Interlocked.Increment(ref _startedRequests);
+            WriteEvent(1, scheme, host, port, pathAndQuery, versionMajor, versionMinor);
```
I use your suggestion of `WriteEvent(eventId: 1, ...)`.

davidsh commented:
/azp run runtime-libraries outerloop

davidsh commented:
/azp run runtime-libraries outerloop

davidsh commented:
CI failures in Outerloop caused by known issue #38205

davidsh commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:
```diff
@@ -57,6 +57,7 @@ public override int Read(Span<byte> buffer)
                     if (_connection == null)
                     {
                         // Fully consumed the response in ReadChunksFromConnectionBuffer.
+                        if (HttpTelemetry.IsEnabled) LogRequestStop();
```
Yes. In general, eventsources including eventcounters can be turned on/off by any listeners (including out-of-proc tools such as PerfView). This is really a known race condition in general across any of our events and counters in .NET Core and ASP.NET. There isn't a solution to this other than to remind listeners to scope the on/off appropriately.
